File: vol2/ch16/LeetCode236.py
class Solution:
    def lowestCommonAncestor(self,root:'TreeNode',p:'TreeNode',q:'TreeNode')->'TreeNode':
        if not root:return None
        if root==p or root==q:return root
        leftans=self.lowestCommonAncestor(root.left,p,q)
        rightans=self.lowestCommonAncestor(root.right,p,q)
        if leftans!=None and rightans!=None:
            return root
        elif leftans!=None:
            return leftans
        elif rightans!=None:
            return rightans
        return None

File: vol2/ch16/LeetCode236.cpp
class Solution {
public:
    TreeNode*lowestCommonAncestor(TreeNode*root,TreeNode*p,TreeNode*q) {
        if(root==NULL) return NULL;
        if(root==p || root==q) return root;
        TreeNode*leftans=lowestCommonAncestor(root->left,p,q);
        TreeNode*rightans=lowestCommonAncestor(root->right,p,q);
        if(leftans!=NULL && rightans!=NULL)
            return root;
        else if(leftans!=NULL)
            return leftans;
        else if(rightans!=NULL)
            return rightans;
        return NULL;
    }
};

